File: java/1.x/picocontainer/tags/picocontainer-1_2/container/src/test/org/picocontainer/defaults/DelegatingPicoContainerTestCase.java
/*****************************************************************************
 * Copyright (c) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 *****************************************************************************/

package org.picocontainer.defaults;

import junit.framework.TestCase;
import org.picocontainer.MutablePicoContainer;
import org.picocontainer.PicoInitializationException;
import org.picocontainer.PicoRegistrationException;
import org.picocontainer.testmodel.DependsOnTouchable;
import org.picocontainer.testmodel.SimpleTouchable;

public class DelegatingPicoContainerTestCase extends TestCase {
    private MutablePicoContainer parent;
    private DefaultPicoContainer child;

    public void setUp() throws PicoRegistrationException, PicoInitializationException {
        parent = new DefaultPicoContainer();
        child = new DefaultPicoContainer(parent);
    }

    public void testChildGetsFromParent() {
        parent.registerComponentImplementation(SimpleTouchable.class);
        child.registerComponentImplementation(DependsOnTouchable.class);
        DependsOnTouchable dependsOnTouchable = (DependsOnTouchable) child.getComponentInstance(DependsOnTouchable.class);

        assertNotNull(dependsOnTouchable);
    }

    public void testParentDoesntGetFromChild() {
        child.registerComponentImplementation(SimpleTouchable.class);
        parent.registerComponentImplementation(DependsOnTouchable.class);
        try {
            parent.getComponentInstance(DependsOnTouchable.class);
            fail();
        } catch (UnsatisfiableDependenciesException e) {
        }
    }

    public void testChildOverridesParent() {
        parent.registerComponentImplementation(SimpleTouchable.class);
        child.registerComponentImplementation(SimpleTouchable.class);

        SimpleTouchable parentTouchable = (SimpleTouchable) parent.getComponentInstance(SimpleTouchable.class);
        SimpleTouchable childTouchable = (SimpleTouchable) child.getComponentInstance(SimpleTouchable.class);
        assertEquals(1, child.getComponentInstances().size());
        assertNotSame(parentTouchable, childTouchable);
    }
}

File: dotnet/picocontainer/src/Test/Extras/DelegatingPicoContainerTestCase.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using System.Threading;
using System.Collections;
using csUnit;

using PicoContainer.Extras;
using PicoContainer.Lifecycle;
using PicoContainer.Defaults;
using PicoContainer.Tests.TestModel;


namespace PicoContainer.Tests.Extras
{

	/// Summary description for DelegatingPicoContainerTestCase.
	/// </summary>
	[TestFixture]
	public class DelegatingPicoContainerTestCase
	{
    private MutablePicoContainer parent;
    private DefaultPicoContainer child;

    public void SetUp() {
      parent = new DefaultPicoContainer();
      child = new DefaultPicoContainer();
      child.AddParent(parent);
    }

    public void testChildGetsFromParent() {
      parent.RegisterComponentImplementation(typeof(SimpleTouchable));
      child.RegisterComponentImplementation(typeof(DependsOnTouchable));
      DependsOnTouchable dependsOnTouchable = (DependsOnTouchable) child.GetComponentInstance(typeof(DependsOnTouchable));

      Assert.NotNull(dependsOnTouchable);
    }

    public void testParentDoesntGetFromChild() {
      child.RegisterComponentImplementation(typeof(SimpleTouchable));
      parent.RegisterComponentImplementation(typeof(DependsOnTouchable));
      try {
        DependsOnTouchable dependsOnTouchable = (DependsOnTouchable) parent.GetComponentInstance(typeof(DependsOnTouchable));
        Assert.Fail();
      } catch (NoSatisfiableConstructorsException) {
      }
    }

    public void testChildOverridesParent() {
      parent.RegisterComponentImplementation(typeof(SimpleTouchable));
      child.RegisterComponentImplementation(typeof(SimpleTouchable));

      SimpleTouchable parentTouchable = (SimpleTouchable) parent.GetComponentInstance(typeof(SimpleTouchable));
      SimpleTouchable childTouchable = (SimpleTouchable) child.GetComponentInstance(typeof(SimpleTouchable));
      Assert.Equals(1, child.ComponentInstances.Count);
      Assert.True(parentTouchable!=childTouchable);
    }

    public void testMulticaster()  {
      parent.RegisterComponentImplementation(typeof(SimpleTouchable));
      child.RegisterComponentImplementation(typeof(DependsOnTouchable));

      Object multicaster = child.GetComponentMulticaster();
      Assert.True(multicaster is Touchable);
    }
  }
}
